proposals/oci-support-updates.md

```diff
@@ -0,0 +1,95 @@
+# Proposal: OCI Support Updates
+
+A couple of the maintainers recently met regarding how to move forward with Helm's experimental OCI support. This documents the conclusions reached.
+
+Note: the existing support for OCI registries is documented [here](https://helm.sh/docs/topics/registries/).
+
+## 1. Implement `Getter` and introduce `Pusher`
+
+As of yet, the code related to registry support has been written standalone (see `internal/experimental/registry`).
+
+The act of downloading a chart from an OCI registry should mimic what is already possible using downloader plugins. In theory, using a protocol prefix such as `oci://` in any place where chart repos are referenced should work just by implementing a `Getter` for OCI.
+
+Currently there is no Helm-specific way to upload chart packages. Using the same model of `Getter`, a new interface called `Pusher` should be introduced, with a single built-in implementation (OCI). This opens the door for plugins to implement their own upload mechanism, and for Helm to add a new top-level upload command such as `helm push`.
+
+
+## 2. Support for provenance files
+
+Helm currently has the ability to verify package signatures, assuming the presence of a file suffixed with `.prov` sitting next to a chart `.tgz` in a repository.
+
+Support for this method of signature validation should be carried over into OCI storage. As the format of the OCI manifest is custom to Helm, Helm can also choose to modify the resulting manifest on upload when a chart is being signed (e.g. `helm push --sign`).
+
+As far as the low-level details, the `.prov` file will simply be stored as a second layer on the manifest. If running `helm pull --verify oci://...`, the second layer (layer at index 1) will be assumed to be the provenance file. The first layer is always the chart `.tgz` itself.
+
+Here is an example of what a manifest might look like with a provenance file attached:
+
+```
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.cncf.helm.config.v1+json",
+    "digest": "sha256:8ec7c0f2f6860037c19b54c3cfbab48d9b4b21b485a93d87b64690fdb68c2111",
+    "size": 117
+  },
```

Comment:
Hey folks,
A few notes:
> And on that note, perhaps `application/vnd.cncf.helm.config.v1+json` is not the correct `mediaType` for an APIVersion 2 Chart.yaml? Should that be `v2`, to track the APIVersion? Or does that introduce complexity without benefit, given that any tool operating on this _can_ just check the APIVersion of the resulting object during or after deserialisation.

Using `v1` or `v3` as it aligns with Helm 3 is a good discussion. Having lived through .NET versioning issues, keeping version numbers aligned, as much as possible = simplicity. Functionally, it's a detail that just requires someone to grok. It's there to enable client tools to know the difference if serialization changes. Such as serializing as `.json` or `.yaml`, or some other additions over time.

> That's not a layer, that's the content type of this object. Oddly enough, it's mentioned as an example in the [OCI Artifact Author's Guide](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-oci-artifact-types) although as `application/vnd.cncf.helm.chart.config.v1+json` rather than `application/vnd.cncf.helm.config.v1+json`.

The OCI artifact doc was written before the Helm work was complete. I'll update to whatever the Helm team decides, including the versioning question above.

As for the config object vs. a layer:
`manifest.config`: More details [here](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-a-unique-artifact-type), but the config object is used to pivot the artifact type. Like a file extension. The value of the config object is optional. More detail [here](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#optional-defining-config-schema)

Reply:
@SteveLasker The chart YAML format also has an APIVersion, it was v1 for Helm v2, and v2 for Helm v3. So that's another version numbering system in the mix. Given that Helm v3 can both generate and process both APIVersion v1 and APIVersion v2 Charts, I would lean more towards tracking the chart APIVersion here than the Helm release version.

I would personally lean towards the OCI mediaType version tracking the APIVersion of the contained Chart. After all, there's no reason I shouldn't be able to upload an APIVersion v1 chart to OCI, and a hosting repository indexer that only knows APIVersion v2 should be able to know it it can't index it.